lib/service/storage/virtualIndex.ts

```diff
@@ -0,0 +1,157 @@
+import Crypto from "crypto";
+
+import { Service } from "../service";
+
+export class VirtualIndex extends Service {
+  public static createEvent = "core:storage:virtualindex:cache:create";
+  public static deleteEvent = "core:storage:virtualindex:cache:delete";
+
+  public static createVirtualIndexEvent =
+    "core:storage:virtualindex:create:after";
+
+  public static deleteVirtualIndexEvent =
+    "core:storage:virtualindex:delete:after";
+
+  constructor() {
+    super("VirtualIndex", global.kuzzle.config.services.storageEngine);
+  }
+
+  public softTenant: Map<string, string> = new Map(); //Key : virtual index, value : real index //TODO rename!
+
+  async init(): Promise<void> {
+    await this.buildCollection();
+    await this.initVirtualTenantList();
+
+    global.kuzzle.onAsk(VirtualIndex.createEvent, (real, virtual) =>
+      this.addInSoftTenantMap(real, virtual)
+    );
+
+    global.kuzzle.onAsk(VirtualIndex.deleteEvent, (info) =>
+      this.removeInSoftTenantMap(info)
+    );
+  }
+
+  addInSoftTenantMap(real, virtual) {
+    this.softTenant.set(virtual, real);
+  }
+
+  removeInSoftTenantMap(notification) {
+    this.softTenant.delete(notification);
+  }
+
+  getRealIndex(name: string): string {
+    if (this.softTenant.has(name)) {
+      return this.softTenant.get(name);
+    }
+    return name;
+  }
+
+  isVirtual(name: string): boolean {
+    return this.softTenant.has(name);
+  }
+
+  getId(index: string, id: string): string {
+    if (this.isVirtual(index)) {
+      return index + id;
+    }
+    return id;
+  }
+
+  randomString(size = 20) {
+    return Crypto.randomBytes(size).toString("base64").slice(0, size);
```

Comment:
> It would be better to generate sequentials IDs so they will be sorted by default by ES directly (and thus in the Admin Console for example)

I'm not  really sure about that since having sequential ID's when not having any form of transaction in ES could result in the same ID generated twice, which could lead to two nodes each creating a document in the same virtual collection and both having the same id which will cause the creation of the document to fail in one of the two nodes.

That said I'm not sure the usage of Crypto is really necessary, there is no sensitive information and crypto is a bit slow but that's not what will bottleneck Kuzzle, so I'm not against it.
Just for reference
```bash
Crypto x 632,951 ops/sec ±0.83% (90 runs sampled)
Without Crypto x 3,115,597 ops/sec ±0.33% (93 runs sampled)
```

**Code without crypto**
```js
  const buf = Buffer.alloc(20);
  for (let i = 0 ; i < 20; i++) {
    buf.writeUInt8((Math.random() * 255)&255, i);
  }
  const id = buf.toString('base64').slice(0, 20);
```

Reply:
done